### PR TITLE
Code smell fixes for BackfillJob

### PR DESCRIPTION
**Pull request by XD-DENG**
Three change items in two parts:

### Part 1
- merge IFs when we can;
- One of the IF condition checks should reuse `STATES_COUNT_AS_RUNNING` which has been already defined;

### Part 2
- Remove unused parameter for intermediate function `_per_task_process()`

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: airflow/jobs/backfill_job.py [hunk 1]
```diff
@@ -425,7 +424,7 @@ def _process_backfill_task_instances(self,  # pylint: disable=too-many-statement
             # determined deadlocked while they are actually
             # waiting for their upstream to finish
             @provide_session
-            def _per_task_process(task, key, ti, session=None):  # pylint: disable=too-many-return-statements
+            def _per_task_process(key, ti, session=None):  # pylint: disable=too-many-return-statements
                 ti.refresh_from_db(lock_for_update=True, session=session)
 
                 task = self.dag.get_task(ti.task_id, include_subdags=True)
```
Comment (XD-DENG) [inline airflow/jobs/backfill_job.py:430]: Clarification for the 2nd part: because we have this line, actually whatever we pass to `task` is ignored

**Review by ashb — APPROVED** [review]
Assuming tests pass LGTM

**Comment by XD-DENG** [comment]
> The PR needs to run all tests because it modifies core of Airflow! Please rebase it to latest master or ask committer to re-run it!

Thanks @ashb . May you please advise any specific thing I should address for this step?

It can also help me understand how I can review/process other contributors' PRs later. Cheers!

**Comment by ashb** [comment]
> > The PR needs to run all tests because it modifies core of Airflow! Please rebase it to latest master or ask committer to re-run it!
> 
> Thanks @ashb . May you please advise any specific thing I should address for this step?
> 
> It can also help me understand how I can review/process other contributors' PRs later. Cheers!

See https://github.com/apache/airflow/pull/12006

**Review by kaxil — APPROVED** [review]

**Comment by potiuk** [comment]
Yeah! @XD-DENG -> it worked :) after rebase you run the full matrix now. I'd love to explain it in a clearer way, but It's not easy :). Happy if you could all help in describing it (and maybe make the comments) a little more obvious what to do!

**Comment by XD-DENG** [comment]
> Yeah! @XD-DENG -> it worked :) after rebase you run the full matrix now. I'd love to explain it in a clearer way, but It's not easy :). Happy if you could all help in describing it (and maybe make the comments) a little more obvious what to do!

Thanks @potiuk . Yes I'm also making use of this PR to explore how the new PR test/review process works. Earlier I tried another option ("re-run workflow" as committer), but seems not working correctly. I expect some minor adjustments would be needed to the process, will share with you once I find any.

**Comment by potiuk** [comment]
> Thanks @potiuk . Yes I'm also making use of this PR to explore how the new PR test/review process works. Earlier I tried another option ("re-run workflow" as committer), but seems not working correctly. I expect some minor adjustments would be needed to the process, will share with you once I find any.

Yep. I am on a lookout for any adjustments and fixes needed :). Happy to get feedback.

**Comment by potiuk** [comment]
@XD-DENG -> the "re-running" problem might be fixed when we merge this one: https://github.com/apache/airflow/pull/12018 . There was a scenario that re-running an old run caused self-cancelling of the Image Build job :) . I've added self-preservation mechanisms to prevent this from happening.

**Pull request merged**